### ISSUE TITLE
[ITensorGPU] [BUG] StackOverflow using similar on CUDA ITensors

### ISSUE THREAD

**Issue by wladiKrin** [ITensor/ITensors.jl#1054]
**Description of bug**

Calling similar on a CUDA ITensor object causes a StackOverflow error.

**Minimal code demonstrating the bug or unexpected behavior**

Running the following example:

```
using ITensors, ITensorGPU
j = Index(10)
phi = cu(randomITensor(j))
similar(phi)
```

leads to the following error output:

```
ERROR: StackOverflowError:
Stacktrace:
 [1] similar(::Type{SimpleTraits.Not{NDTensors.IsWrappedArray{DataType}}}, ::Type, ::Type, ::Type, ::Vararg{Any}) (repeats 13472 times)
   @ NDTensors ~/.julia/packages/SimpleTraits/l1ZsK/src/SimpleTraits.jl:331
 [2] similar(D::NDTensors.Dense{Float64, CUDA.CuArray{Float64, 1, CUDA.Mem.DeviceBuffer}})
   @ NDTensors ~/.julia/packages/NDTensors/FHyfI/src/dense/dense.jl:75
 [3] similar(T::NDTensors.DenseTensor{Float64, 1, Tuple{Index{Int64}}, NDTensors.Dense{Float64, CUDA.CuArray{Float64, 1, CUDA.Mem.DeviceBuffer}}})
   @ NDTensors ~/.julia/packages/NDTensors/FHyfI/src/tensor.jl:174
 [4] similar(::ITensor)
   @ ITensors ~/.julia/packages/ITensors/AXb8G/src/itensor.jl:1699
 [5] top-level scope
   @ REPL[4]:1
 [6] top-level scope
   @ ~/.julia/packages/CUDA/BbliS/src/initialization.jl:52
```

**Possible fix**

The error is caused by a call to the wrong `similar` function.
Overloading fixes the error:
```
NDTensors.similar(D::NDTensors.Dense) = NDTensors.Dense(Base.similar(NDTensors.data(D)))
```

**Version information**

 - Output from `versioninfo()`:
```julia
julia> versioninfo()

Julia Version 1.8.5
Commit 17cfb8e65ea (2023-01-08 06:45 UTC)
Platform Info:
  OS: Linux (x86_64-linux-gnu)
  CPU: 40 × Intel(R) Xeon(R) Silver 4210 CPU @ 2.20GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-13.0.1 (ORCJIT, cascadelake)
  Threads: 1 on 40 virtual cores
```
 - Output from `using Pkg; Pkg.status("ITensors")`:
```julia
julia> using Pkg; Pkg.status("ITensors")
ITensors v0.3.24
ITensorGPU v0.1.0
```

**Comment by mtfishman** [comment]
Thanks, this was previously reported in https://github.com/ITensor/ITensorTDVP.jl/issues/51 and should be fixed by #1051. That will be registered soon.

**Comment by mtfishman** [comment]
This should be fixed by #1051, please report back if you think it isn't fixed.
